iroh/src/discovery/pkarr/dht.rs

```diff
@@ -395,11 +313,102 @@ impl Discovery for DhtDiscovery {
         _endpoint: Endpoint,
         node_id: NodeId,
     ) -> Option<Boxed<anyhow::Result<DiscoveryItem>>> {
-        let this = self.clone();
         let pkarr_public_key =
             pkarr::PublicKey::try_from(node_id.as_bytes()).expect("valid public key");
         tracing::info!("resolving {} as {}", node_id, pkarr_public_key.to_z32());
-        Some(Gen::new(|co| async move { this.gen_resolve(node_id, co).await }).boxed())
+
+        let mut stream = futures_buffered::FuturesUnorderedBounded::new(2);
+        if self.0.pkarr_relay.is_some() {
+            tracing::info!(
+                "resolving {} from relay {:?}",
+                pkarr_public_key.to_z32(),
+                self.0.relay_url
+            );
+            let key = pkarr_public_key.clone();
+            let discovery = self.0.clone();
+            stream.push(
+                async move {
+                    let maybe_packet = discovery
+                        .pkarr_relay
+                        .as_ref()
+                        .expect("checked")
+                        .resolve(&key)
+                        .await;
+                    match maybe_packet {
+                        Ok(Some(signed_packet)) => {
+                            match NodeInfo::from_pkarr_signed_packet(&signed_packet) {
+                                Ok(node_info) => {
+                                    let node_addr: NodeAddr = node_info.into();
+
+                                    tracing::info!(
+                                        "discovered node info from relay {:?}",
+                                        node_addr
+                                    );
+                                    Some(Ok(DiscoveryItem {
+                                        node_addr,
+                                        provenance: "relay",
+                                        last_updated: None,
+                                    }))
+                                }
+                                Err(_err) => {
+                                    tracing::debug!("failed to parse signed packet as node info");
+                                    None
+                                }
+                            }
+                        }
+                        Ok(None) => {
+                            tracing::debug!("no signed packet found in relay");
+                            None
+                        }
+                        Err(err) => {
+                            tracing::debug!("failed to get signed packet from relay: {}", err);
+                            Some(Err(err.into()))
+                        }
+                    }
+                }
+                .boxed(),
```

Comment:
Is it possible to keep having these two extracted out into functions? I think that'd make this a lot more readable.

Reply:
done